File: backend/src/pipeline/submodels.ts
/**
 * AI helpers used only when deterministic parsing isn't enough:
 *  - `structureResume`: LLM fallback that structures a resume from raw text
 *    (used only when the parser's `parse_quality` is weak).
 *  - `classifyCertificates`: vision classification of certificate-candidate
 *    images (gated + capped by the orchestrator; icons/logos never reach here).
 * Each takes an injected `call` (Gemini) so it stays testable.
 */
import type { GeminiRequest } from "../llm/geminiClient.js";
import { buildResumePrompt, RESUME_SCHEMA, IMAGE_PROMPT, IMAGE_SCHEMA } from "../llm/prompts.js";

export type GeminiCaller = (req: GeminiRequest) => Promise<any>;
export type ImageLoader = (path: string) => Promise<{ mimeType: string; data: Buffer }>;

export interface RawImage {
  index: number;
  path: string;
}

export interface ClassifiedImage {
  index: number;
  imageType: string;
  details: unknown;
}

/** Fallback: raw text + explicit links -> structured resume JSON (RESUME_SCHEMA). */
export async function structureResume(
  pipelineA: { text: string; links: unknown[] },
  call: GeminiCaller,
): Promise<any> {
  return call({ prompt: buildResumePrompt(pipelineA), schema: RESUME_SCHEMA });
}

/** Vision-classify certificate-candidate images and extract their details. */
export async function classifyCertificates(
  images: RawImage[],
  call: GeminiCaller,
  loadImage: ImageLoader,
): Promise<ClassifiedImage[]> {
  const out: ClassifiedImage[] = [];
  for (const img of images) {
    const { mimeType, data } = await loadImage(img.path);
    const res = await call({
      prompt: IMAGE_PROMPT,
      schema: IMAGE_SCHEMA,
      images: [{ mimeType, data }],
    });
    out.push({ index: img.index, imageType: res.imageType, details: res.details ?? null });
  }
  return out;
}

File: backend-py/app/pipeline/submodels.py
"""AI helpers used only when deterministic parsing isn't enough (port of submodels.ts).

- structure_resume: LLM fallback that structures a resume from raw text (used only
  when the parser's parse_quality is weak).
- classify_certificates: vision classification of certificate-candidate images
  (gated + capped by the orchestrator; icons/logos never reach here).
Each takes an injected `call` (Gemini) so it stays isolated.
"""

from __future__ import annotations

from typing import Any, Awaitable, Callable

from app.llm.gemini import GeminiImage, GeminiRequest
from app.llm.prompts import (
    IMAGE_PROMPT,
    IMAGE_SCHEMA,
    RESUME_SCHEMA,
    build_resume_prompt,
)

GeminiCaller = Callable[[GeminiRequest], Awaitable[Any]]
ImageLoader = Callable[[str], Awaitable[GeminiImage]]


async def structure_resume(pipeline_a: dict, call: GeminiCaller) -> Any:
    """Fallback: raw text + explicit links -> structured resume JSON (RESUME_SCHEMA)."""
    return await call(
        GeminiRequest(prompt=build_resume_prompt(pipeline_a), schema=RESUME_SCHEMA)
    )


async def classify_certificates(
    images: list[dict],
    call: GeminiCaller,
    load_image: ImageLoader,
) -> list[dict]:
    """Vision-classify certificate-candidate images and extract their details."""
    out: list[dict] = []
    for img in images:
        loaded = await load_image(img["path"])
        res = await call(
            GeminiRequest(prompt=IMAGE_PROMPT, schema=IMAGE_SCHEMA, images=[loaded])
        )
        out.append(
            {
                "index": img["index"],
                "imageType": res.get("imageType"),
                "details": res.get("details"),
            }
        )
    return out
